pkg/timelock/timelock.go

```diff
@@ -132,63 +132,81 @@ func (tw *Worker) Listen(ctx context.Context) error {
 		return err
 	}
 
+	tw.startLog()
+
+	// Handle OS signals to properly stop/kill the process.
 	stopCh := make(chan string)
-	logCh := make(chan types.Log)
-	sigCh := make(chan os.Signal, 1)
-	signal.Notify(sigCh, syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM)
+	go handleOSSignal(stopCh)
 
+	// Update timelock-worker default scheduler delay.
 	tw.updateSchedulerDelay(time.Duration(tw.pollPeriod) * time.Second)
+
+	// Run the scheduler to add/del operations in a thread-safe way.
 	go tw.runScheduler(ctx)
-	tw.startLog()
 
-	// Shutdown gracefully based on OS interrupts.
-	go func() {
-		for {
-			handleOSSignal(<-sigCh, stopCh)
-		}
-	}()
+	// Initialize the subscription.
+	logCh := make(chan types.Log)
+	if err := tw.subscribeAndProcessLogs(ctx, logCh, stopCh); err != nil {
+		tw.logger.Error().Err(err).Msg("failed to subscribe and process logs.")
+		return err
+	}
+
+	defer close(stopCh)
+	defer close(logCh)
+	defer ctx.Done()
+	defer tw.dumpOperationStore(time.Now)
+
```

Comment:
none of this cleanup happens on error from `subscribeAndProcessLogs`

Reply:
they would be executed:
- In case of a sub.Err() on MainLoop : MainLoop breaks -> wg.Done() is executed -> return to Listen() -> rest of deferred calls.
- In case of returning an error in FilterLogs or SubscribeFilterLogs: deferred sub.Unsuscribe() call -> return err -> log "failed to subscribe and process logs." -> rest of deferred calls -> return err.

Reply:
These defer calls are made after line 151.  Since `tw.subscribeAndProcessLogs` is a blocking call these defer statements are never called within Listen.  So if there is an error in `FilterLogs` then you log the error and return err from Listen.  Listen is now done and no defer calls are made.